src/AggregateFunctions/AggregateFunctionGroupConcat.h

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+#ifndef DB_GROUP_CONCAT_H
+#define DB_GROUP_CONCAT_H
+
+#include <AggregateFunctions/IAggregateFunction.h>
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <AggregateFunctions/FactoryHelpers.h>
+#include <Core/ServerSettings.h>
+#include <Common/ArenaAllocator.h>
+#include <DataTypes/IDataType.h>
+#include <DataTypes/DataTypeString.h>
+#include <IO/WriteHelpers.h>
+#include <IO/ReadHelpers.h>
+
+namespace DB
+{
+struct Settings;
+
+struct GroupConcatDataBase
+{
+    UInt64 data_size = 0;
+    UInt64 allocated_size = 0;
+    char * data = nullptr;
+
+    void checkAndUpdateSize(UInt64 add, Arena * arena);
+    void insertChar(const char * str, UInt64 str_size, Arena * arena);
+    void insert(const IColumn * column, const SerializationPtr & serialization, size_t row_num, Arena * arena);
+};
+
+template <bool has_limit>
+struct GroupConcatData : public GroupConcatDataBase
```

Comment:
Am I right that after this PR GroupConcatData<false> and GroupConcatData<true> are going to be identical? In this case we do not need neither template nor the Base class.

Reply:
`<bool has_limit>` determines whether we have limitation for output for our query or not:

``` C++
    if constexpr (has_limit)
        if (cur_data.num_rows >= limit)
            return;
```
            
so they are not identical in it's implementation and usage: `has_limit` is true when the second parameter (limit) is used: https://clickhouse.com/docs/en/sql-reference/aggregate-functions/reference/groupconcat

Reply:
```
  if constexpr (has_limit)
        if (cur_data.num_rows >= limit)
            return;
```

This rows are from GroupConcatImpl. I wonder if there are any differences in GroupConcatData

Reply:
yes, you are right, I will remove templates from GroupConcatData